stdlib/dynarray.mli

```diff
@@ -479,3 +479,121 @@ val reset : 'a t -> unit
+comparison function.
+
+{[
+(* We present our priority queues as a functor
+   parametrized on the comparison function. *)
+module Heap (Elem : Map.OrderedType) : sig
+  type t
+  val create : unit -> t
+  val add : t -> Elem.t -> unit
+  val pop_min : t -> Elem.t option
+end = struct
+
+  (* Our priority queues are implemented using the standard "min heap"
+     data structure, a dynamic array representing a binary tree. *)
+  type t = Elem.t Dynarray.t let create = Dynarray.create
+
+ (* The node of index [i] has as children the nodes of index [2 * i + 1]
+    and [2 * i + 2] -- if they are valid indices in the dynarray. *)
+  let left_child i = 2 * i + 1
+  let right_child i = 2 * i + 2
+  let parent_node i = (i - 1) / 2
+
+  (* We use indexing operators for convenient notations. *)
+  let ( .%() ) = Dynarray.get
+  let ( .%()<- ) = Dynarray.set
+
+  (* Auxiliary functions to compare and swap two elements
+     in the dynamic array. *)
+  let order h i j =
+    Elem.compare h.%(i) h.%(j)
+
+  let swap h i j =
+    let v = h.%(i) in
+    h.%(i) <- h.%(j);
+    h.%(j) <- v
+
+  (* We say that a heap respects the "heap ordering" if the value of
+     each node is smaller than the value of its children. The
+     algorithm manipulates arrays that respect the heap algorithm,
+     except for one node whose value may be too small or too large.
+
+     The auxiliary functions [heap_up] and [heap_down] take
+     such a misplaced value, and move it "up" (respectively: "down")
+     the tree by permuting it with its parent value (respectively:
+     a children's value) until the heap ordering is restored. *)
+
+  let rec heap_up h i =
+    if i = 0 then () else
+    let parent = parent_node i in
+    if order h i parent < 0 then
+      (swap h i parent; heap_up h parent)
+
+  and heap_down h ~len i =
+    let left, right = 2 * i + 1, 2 * i + 2 in
+    if left >= len then () (* no child, stop *) else
+    let smallest =
+      if right >= len then left (* no right child *) else
+      if order h left right < 0 then left else right
+    in
+    if order h i smallest > 0 then
+      (swap h i smallest; heap_down h ~len smallest)
+
+  let add h s =
+    let i = Dynarray.length h in
+    Dynarray.add_last h s;
+    heap_up h i
+
+  let pop_min h =
+    if Dynarray.is_empty h then None
+    else begin
+      (* Standard trick: swap the 'best' value at index 0
+         with the last value of the array. *)
+      let last = Dynarray.length h - 1 in
+      swap h 0 last;
+      (* At this point [pop_last] returns the 'best' value,
+         and leaves a heap with one misplaced element at position 0. *)
+      let best = Dynarray.pop_last h in
+      (* Restore the heap ordering -- does nothing if the heap is empty. *)
+      heap_down h ~len:last 0;
+      Some best
+    end
+end
+]}
+
+The code above is inspired from an ISC-licensed real-world
+implementation by Daniel Bünzli. The real code includes logic to free
```

Comment:
Thanks for the relicensing pledge which is making my life slightly easier[^1]. I am slightly frustrated by your no-crediting recommendation (it's not my way of doing things) but followed your preference.

[^1]: In practice people reusing code copy-pasted from the manual should potentially be *more* worried now that its license is even less clear. Oh well.

Reply:
In general I have the sample code of the docs of my libraries extracted somewhere to make sure it compiles (sadly a manual process because https://github.com/ocaml/odoc/pull/303 has been lingering for years). In that file I indicate that the code is in the public domain (yes I know about its inexistence in some galaxies). 

What would upstream think about making a statement somewhere that sample code from the stdlib docs is in the public domain/[CC0](https://creativecommons.org/share-your-work/public-domain/cc0/) ?

Reply:
Yes, the commit that introduces this example also duplicates it in the testsuite. I had an attribution comment there which I also removed in an overly broad interpretation of your remark, I'll put it back. I should also add cross-references to the copies in both directions to reduce the chances that they desynchronize in the future. (I do mirror changes by hand using `meld` and it is indeed slightly annoying.)

Reply:
> What would upstream think about making a statement somewhere that sample code from the stdlib docs is in the public domain/[CC0](https://creativecommons.org/share-your-work/public-domain/cc0/) ?

Would you open a separate issue do discuss this? I agree that it is a good idea, but I don't have the energy to converge on this in the context of the present PR.

Reply:
Issue https://github.com/ocaml/ocaml/issues/12095